tao_compiler/mlir/disc/transforms/disc_shape_optimization_utils.cc

```diff
@@ -594,6 +665,72 @@ SymbolicDimMgr::getOrCreateSymbolicDimsForRankedValue(Value value) {
   return symbols;
 }
 
+SymbolicDimOp SymbolicDimMgr::cloneSymbol(SymbolicDimOp symbol) {
+  // early return for static shape symbol
+  if (!symbol.isDynamic()) return symbol;
+  SymbolicDimOp newSymbol = newSymbolicDim();
+
+  // copied all attributes expect the name.
+  for (const NamedAttribute& attr : symbol->getAttrs()) {
+    StringRef name = attr.getName().strref();
+    if (name == "sym_name") continue;
+    newSymbol->setAttr(name, attr.getValue());
+  }
+  return newSymbol;
+}
+
+LogicalResult SymbolicDimMgr::cloneSymbolGroup(
+    const DenseSet<SymbolicDimOp>& symbols,
+    DenseMap<SymbolicDimOp, SymbolicDimOp>& mapping) {
+  DenseMap<SymbolicDimOp, SmallVector<SymbolicDimOp>> symbol2Root;
+  for (SymbolicDimOp symbol : symbols) {
+    SymbolicDimOp newSymbol = cloneSymbol(symbol);
+    mapping[symbol] = newSymbol;
+    SymbolicDimOp root = getRootSymbolicDim(symbol);
+    symbol2Root[root].push_back(newSymbol);
+  }
+
+  // copy dim equality predicate.
+  for (auto& it : symbol2Root) {
+    if (it.second.size() <= 1) continue;
+    for (SymbolicDimOp symbol : it.second)
+      if (failed(mapSymbolicDimEqual(symbol, it.second[0]))) return failure();
```

Comment:
why don't set root of all symbol in `it.second` to `root` on L690, but to each other?

Reply:
we want to build equality relationship between the new symbols. However, we can not guarantee that the images of other symbols in the some group are created when we visit the current symbol and create image for it. Thus, we use a more safe method.